File: CSharpCodeBase/aiactions/ai_following.lua
local Class = require 'utils.hump.class'
local Timer = require 'utils.hump.timer'
local AINode = require 'utils.ailib.ainode'

local AI_Following = Class({__includes=AINode, init = function (self, entity)
  AINode.init(self, entity)
  self.minRadius = self.entity.config:Get('minDistanceProsecution')
  self.maxRadius = self.entity.config:Get('maxDistanceProsecution')
end})

function AI_Following:Visit()
  self.target = self.entity.goalProsecution
  if self.target then
    if self.status == NODE_READY then
      self.status = NODE_RUNNING
      self:Moving()
    elseif self.status == NODE_RUNNING then
      if not self.target.interactable then
        self.status = NODE_FAILURE
        return
      end
      if not self.entity.mover:IsHaveGoal() then
        self.status = NODE_SUCCESS
      end
    end
  end
  return self.status
end

function AI_Following:Moving()
  self.entity.mover:ResetSpeed()
  self.entity.mover:LookAt(self.target:GetPosition())
  self.entity.mover:SetGoal(self.target, nil, math.random(self.minRadius, self.maxRadius))
end

return AI_Following

File: CSharpCodeBase/aiactions/ai_following.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class ai_following :ainode{
 public void init (self, entity);
   AINode.init(self, entity);
   this.minRadius = this.entity.config:Get("minDistanceProsecution");
   this.maxRadius = this.entity.config:Get("maxDistanceProsecution");
 }})
 public void Visit(){
   this.target = this.entity.goalProsecution;
   if(this.target  ){ 
     if(this.status == NODE_READY  ){ 
       this.status = NODE_RUNNING;
       self:Moving();
     }else{if this.status == NODE_RUNNING  ){ 
       if(not this.target.interactable  ){ 
         this.status = NODE_FAILURE;
         return;
       }
       if(not this.entity.mover:IsHaveGoal()  ){ 
         this.status = NODE_SUCCESS;
       }
     }
   }
   return this.status;
 }
 public void Moving(){
   this.entity.mover:ResetSpeed();
   this.entity.mover:LookAt(this.target:GetPosition());
   this.entity.mover:SetGoal(this.target, null, math.random(this.minRadius, this.maxRadius));
 }
 }}